### PR TITLE
Make rclpy.try_shutdown() behavior to follow rclpy.shutdown() more closely

### PR DESCRIPTION
**Pull request by ivanpauno**
Needed by https://github.com/ros2/ros2cli/pull/683.

I'm not sure if this is the ideal approach, maybe init()/shutdown() are doing too much (?).
I'm open to alternatives.

### CODE REVIEW

File: rclpy/rclpy/utilities.py [hunk 1]
```diff
@@ -60,9 +60,19 @@ def shutdown(*, context=None):
...
-        context = get_default_context()
-    return context.try_shutdown()
+        # Replace the default context with a new one if shutdown was successful
+        # or if the context was already shutdown.
+        with g_context_lock:
+            if g_default_context is None:
+                g_default_context = Context()
+            g_default_context.try_shutdown()
+            if not g_default_context.ok():
+                g_default_context = None
```
Comment (sloretz) [inline rclpy/rclpy/utilities.py:73]: I wish the default context registered an on_shutdown callback to manage the global variable too.

File: rclpy/rclpy/__init__.py [hunk 2]
```diff
@@ -129,6 +129,23 @@ def shutdown(*, context: Context = None, uninstall_handlers: Optional[bool] = No
...
+
+    :param context: The context to invalidate. If ``None``, then the default context is used
+        (see :func:`.get_default_context`).
+    """
+    global __executor
+    _try_shutdown(context=context)
+    if not ok():
+        if __executor is not None:
+            __executor.shutdown()
+            __executor = None
```
Comment (sloretz) [inline rclpy/rclpy/__init__.py:146]: I wish the default executor registered an `on_shutdown()` callback with the default Context and handled itself that way. Would you be willing to try that? If not, I'd approve this change as it does seem to make the two shutdown functions more alike.

File: rclpy/rclpy/utilities.py (hunk 1 above)
Comment (ivanpauno) [reply rclpy/rclpy/utilities.py:73]: Currently all the work is done with the default context lock taken.
If we register an on_shutdown() callback I don't think that would work.

Maybe it's possible to do that with some extra trick.

File: rclpy/rclpy/__init__.py (hunk 2 above)
Comment (ivanpauno) [reply rclpy/rclpy/__init__.py]: That seems to be working correctly, see https://github.com/ros2/rclpy/pull/868/commits/cb23bd7827a6e69c7348021d53dc1877bd9fc705

**Comment by ivanpauno** [comment]
I had one extra issue related to https://github.com/ros2/rclpy/pull/868#discussion_r774045161 after doing some further testing.
The `reset_executor()` closure I was passing to `on_shutdown()` doesn't have any owner.
As `Context` only retains weak ownership of the callbacks being passed it was sometimes not being called.

https://github.com/ros2/rclpy/pull/868/commits/d65b4871a3122a927703c8cc94f1ccc2dae628a8 fixes that.
After that change, we retaing week ownership of methods and shared ownership of free functions or closures.
I think that's intuitive.
I could also see the case of always retaining shared ownership and the user explictily passing a weak object when they need, but that would be an unexpected change (particularly for methods).

File: rclpy/rclpy/context.py [hunk 3]
```diff
@@ -127,7 +127,7 @@ def on_shutdown(self, callback: Callable[[], None]):
                 if ismethod(callback):
                     self._callbacks.append(weakref.WeakMethod(callback, self._remove_callback))
                 else:
-                    self._callbacks.append(weakref.ref(callback, self._remove_callback))
+                    self._callbacks.append(callback)
```
Comment (sloretz) [inline rclpy/rclpy/context.py:130]: > I could also see the case of always retaining shared ownership and the user explictily passing a weak object when they need that, but that would be an unexpected change (particularly for methods).

Oh hmm. This is interesting. I don't see a better way, and it does seem like the behavior would be intuitive.

**Review by sloretz — APPROVED** [review]

**Comment by ivanpauno** [comment]
CI for this PR alone (I will rerun PR for https://github.com/ros2/ros2cli/pull/683 after merging this one):

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15830/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10506/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13494/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16114/)

**Pull request merged**